File: lint-arwaky/archive/capabilities/mcp_tool_schema_checker.py
"""McpToolSchemaChecker — AES025 for MCP tools JSON Schema compliance.

AES025 MCP_TOOL_SCHEMA_VIOLATION:
MCP tools must declare valid JSON Schema for their input parameters.
Detects: missing schema, empty schema, invalid JSON Schema syntax,
missing descriptions, missing required fields array when properties exist.
"""
from __future__ import annotations

import re
import ast

from ..taxonomy import (
    ColumnNumber,
    ErrorCode,
    FilePath,
    FilePathList,
    LintMessage,
    LintResult,
    LintResultList,
    LineNumber,
    Severity,
)


# JSON Schema draft-07/2020-12 required keywords
_JSON_SCHEMA_KEYWORDS = frozenset({
    "type", "properties", "required", "items",
    "additionalProperties", "description", "title",
    "enum", "const", "default", "minimum", "maximum",
    "minLength", "maxLength", "pattern", "format",
    "anyOf", "oneOf", "allOf", "not",
    "$ref", "$defs", "$schema",
})

# Patterns that indicate a tool registration (FastMCP, stdio MCP, etc.)
_TOOL_DECORATOR_PATTERNS = [
    re.compile(r"@\w+\.tool\s*\("),       # @mcp.tool(...)
    re.compile(r"@\w+\.tool\s*$"),        # @mcp.tool
    re.compile(r"server\.add_tool\b"),    # server.add_tool(...)
    re.compile(r"register_tool\b"),       # register_tool(...)
]

_JSON_SCHEMA_TYPE_VALUES = frozenset({
    "string", "number", "integer", "boolean",
    "array", "object", "null",
})


class McpToolSchemaChecker:
    """AES025 — Validate MCP tool input/output schemas."""

    def check_mcp_tool_schema(
        self,
        analyzer,
        files: FilePathList,
        root_dir: FilePath,
        results: LintResultList,
    ) -> None:
        """Scan all files for MCP tool schema violations."""
        for f in files:
            if not str(f).endswith(".py"):
                continue
            self._check_file(f, results)

    def _check_file(self, f: FilePath, results: LintResultList) -> None:
        """Parse a Python file and find tool registration patterns."""
        try:
            content = open(f.value, "r", encoding="utf-8").read()
            tree = ast.parse(content, filename=f.value)
        except (SyntaxError, UnicodeDecodeError, OSError):
            return

        for node in ast.walk(tree):
            # Check decorated tool functions
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                for decorator in node.decorator_list:
                    if self._is_tool_decorator(decorator, content):
                        self._check_tool_schema(node, f, results)

    # ------------------------------------------------------------------
    # Detection
    # ------------------------------------------------------------------
    def _is_tool_decorator(self, decorator: ast.expr, content: str) -> bool:
        """Determine if a decorator node represents an MCP tool registration."""
        if isinstance(decorator, ast.Call):
            if isinstance(decorator.func, ast.Attribute):
                return decorator.func.attr == "tool"
        if isinstance(decorator, ast.Attribute):
            return decorator.attr == "tool"
        return False

    # ------------------------------------------------------------------
    # Schema validation
    # ------------------------------------------------------------------
    def _check_tool_schema(
        self, func: ast.FunctionDef | ast.AsyncFunctionDef,
        f: FilePath, results: LintResultList,
    ) -> None:
        """Validate the schema associated with a tool function."""
        func_name = func.name
        self._check_docstring(func, f, func_name, results)
        self._check_parameter_types(func, f, func_name, results)
        self._check_explicit_schemas(func, f, func_name, results)

    def _check_docstring(
        self, func: ast.FunctionDef | ast.AsyncFunctionDef,
        f: FilePath, func_name: str, results: LintResultList,
    ) -> None:
        """Tools must have a docstring — this becomes the tool description in tools/list."""
        docstring = ast.get_docstring(func)
        if not docstring or len(docstring.strip()) < 10:
            results.values.append(
                LintResult(
                    code=ErrorCode(code="AES025"),
                    message=LintMessage(
                        value=(
                            f"AES025 MCP_TOOL_SCHEMA_VIOLATION: MCP tool "
                            f"'{func_name}' is missing a descriptive docstring.\n"
                            f"WHY? The docstring becomes the tool description "
                            f"in tools/list response — models use it for routing.\n"
                            f"FIX: Add a docstring describing what this tool does, "
                            f"its inputs, and expected output."
                        )
                    ),
                    severity=Severity.CRITICAL,
                    file=f,
                    line=LineNumber(value=func.lineno),
                    column=ColumnNumber(value=1),
                )
            )

    def _check_parameter_types(
        self, func: ast.FunctionDef | ast.AsyncFunctionDef,
        f: FilePath, func_name: str, results: LintResultList,
    ) -> None:
        """All non-self parameters on a tool function must have type annotations."""
        for arg in func.args.args:
            if arg.arg == "self" or arg.arg == "ctx":
                continue
            if not arg.annotation:
                results.values.append(
                    LintResult(
                        code=ErrorCode(code="AES025"),
                        message=LintMessage(
                            value=(
                                f"AES025 MCP_TOOL_SCHEMA_VIOLATION: MCP tool "
                                f"'{func_name}' parameter '{arg.arg}' lacks a type annotation.\n"
                                f"WHY? Untyped parameters cannot be mapped to JSON Schema "
                                f"in the tools/list schema — models won't know the input format.\n"
                                f"FIX: Add a type annotation (e.g., str, int, FilePath, or a "
                                f"Pydantic model)."
                            )
                        ),
                        severity=Severity.CRITICAL,
                        file=f,
                        line=LineNumber(value=arg.lineno or func.lineno),
                        column=ColumnNumber(value=1),
                    )
                )

    def _check_explicit_schemas(
        self, func: ast.FunctionDef | ast.AsyncFunctionDef,
        f: FilePath, func_name: str, results: LintResultList,
    ) -> None:
        """Check for explicit JSON Schema dicts passed to the tool."""
        for decorator in func.decorator_list:
            if isinstance(decorator, ast.Call):
                for keyword in decorator.keywords:
                    if keyword.arg in ("parameters", "input_schema", "schema"):
                        self._validate_schema_value(keyword.value, f, func_name, results)

    def _validate_schema_value(
        self, node: ast.expr, f: FilePath, func_name: str, results: LintResultList,
    ) -> None:
        """If an explicit schema dict is found, validate it."""
        if not isinstance(node, ast.Dict):
            return

        keys = _collect_dict_string_keys(node)
        violations: list[str] = []
        self._check_required_keywords(keys, violations)
        self._check_property_descriptions(node, keys, violations)
        self._check_type_values(node, keys, violations)

        if violations:
            self._report_aes025(func_name, node, violations, f, results)

    # -- Sub-checks for _validate_schema_value -----------------------------------

    def _check_required_keywords(self, keys: list[str], violations: list[str]) -> None:
        """Verify that required JSON Schema keywords are present."""
        if "type" not in keys:
            violations.append("Schema missing 'type' keyword")
            return
        if "properties" not in keys:
            violations.append("Schema has 'type' but no 'properties'")

    def _check_property_descriptions(
        self, node: ast.Dict, keys: list[str], violations: list[str],
    ) -> None:
        """Ensure each property in the schema has a description."""
        if "properties" not in keys:
            return
        props_value = node.values[keys.index("properties")]
        if not isinstance(props_value, ast.Dict):
            return
        for pkey, pval in zip(props_value.keys, props_value.values):
            if pkey is None:
                continue
            _validate_property_has_description(pkey, pval, violations)

    def _check_type_values(
        self, node: ast.Dict, keys: list[str], violations: list[str],
    ) -> None:
        """Validate that 'type' values are valid JSON Schema types."""
        if "type" not in keys:
            return
        type_node = node.values[keys.index("type")]
        if isinstance(type_node, ast.Constant) and type_node.value not in _JSON_SCHEMA_TYPE_VALUES:
            type_str = str(type_node.value) if isinstance(type_node.value, str) else repr(type_node.value)
            violations.append(f"Schema type='{type_str}' is not a valid JSON Schema type")

    def _report_aes025(
        self, func_name: str, node: ast.expr,
        violations: list[str], f: FilePath, results: LintResultList,
    ) -> None:
        """Append a single AES025 result to the results list."""
        detail = "\n".join(f"  - {v}" for v in violations)
        results.values.append(
            LintResult(
                code=ErrorCode(code="AES025"),
                message=LintMessage(
                    value=(
                        f"AES025 MCP_TOOL_SCHEMA_VIOLATION: MCP tool "
                        f"'{func_name}' has an invalid JSON Schema:\n{detail}\n"
                        f"WHY? MCP tools must declare valid JSON Schema so "
                        f"LLM clients can validate input before tool calls.\n"
                        f"FIX: Use a Pydantic BaseModel for tool parameters "
                        f"or provide a valid dict with 'type' and 'properties' keys."
                    )
                ),
                severity=Severity.CRITICAL,
                file=f,
                line=LineNumber(value=node.lineno or 1),
                column=ColumnNumber(value=1),
            )
        )


# --- module-level helpers ------------------------------------------------------

def _collect_dict_string_keys(node: ast.Dict) -> list[str]:
    """Extract string keys from an ast.Dict node."""
    return [
        k.value for k in node.keys
        if isinstance(k, ast.Constant) and isinstance(k.value, str)
    ]


def _validate_property_has_description(pkey: ast.expr, pval: ast.expr, violations: list[str]) -> None:
    """Flag a property that lacks a 'description' sub-key in its schema."""
    if isinstance(pkey, ast.Constant) and isinstance(pkey.value, str) and isinstance(pval, ast.Dict):
        sub_keys = _collect_dict_string_keys(pval)
        if "description" not in sub_keys:
            violations.append(f"Property '{pkey.value}' missing description in schema")

File: lint-arwaky/src-rust/capabilities/mcp_tool_schema_checker.rs
// mcp_tool_schema_checker — AES025 for MCP tools JSON Schema compliance.
//
// AES025 MCP_TOOL_SCHEMA_VIOLATION:
// MCP tools must declare valid JSON Schema for their input parameters.
// Detects: missing schema, empty schema, invalid JSON Schema syntax,
// missing descriptions, missing required fields array when properties exist.

use crate::taxonomy::*;
use once_cell::sync::Lazy;
use regex::Regex;

// JSON Schema draft-07/2020-12 required keywords
static _JSON_SCHEMA_KEYWORDS: Lazy<std::collections::HashSet<&'static str>> = Lazy::new(|| {
    [
        "type", "properties", "required", "items",
        "additionalProperties", "description", "title",
        "enum", "const", "default", "minimum", "maximum",
        "minLength", "maxLength", "pattern", "format",
        "anyOf", "oneOf", "allOf", "not",
        "$ref", "$defs", "$schema",
    ]
    .iter()
    .copied()
    .collect()
});

// Patterns that indicate a tool registration (FastMCP, stdio MCP, etc.)
static TOOL_DECORATOR_PATTERNS: Lazy<Vec<Regex>> = Lazy::new(|| {
    vec![
        Regex::new(r"@\w+\.tool\s*\(").unwrap(),   // @mcp.tool(...)
        Regex::new(r"@\w+\.tool\s*$").unwrap(),     // @mcp.tool
        Regex::new(r"server\.add_tool\b").unwrap(),  // server.add_tool(...)
        Regex::new(r"register_tool\b").unwrap(),     // register_tool(...)
    ]
});

static JSON_SCHEMA_TYPE_VALUES: Lazy<std::collections::HashSet<&'static str>> = Lazy::new(|| {
    [
        "string", "number", "integer", "boolean",
        "array", "object", "null",
    ]
    .iter()
    .copied()
    .collect()
});

// Regex: captures a function definition line
static FUNC_DEF_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^(?:async\s+)?def\s+(\w+)\s*\(([^)]*)\)").unwrap()
});

// Regex: captures decorator lines
static DECORATOR_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^\s*@(.+)$").unwrap()
});

// Regex: triple-quoted docstring
static DOCSTRING_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"^\s*(?:"""[\s\S]*?"""|'''[\s\S]*?''')"#).unwrap()
});

// Regex: parameter with type annotation
static PARAM_ANNOTATED_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(\w+)\s*:\s*\S+").unwrap()
});

// Regex: extract keyword arguments from decorator call
static KWARG_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(\w+)\s*=\s*").unwrap()
});

/// AES025 — Validate MCP tool input/output schemas.
pub struct McpToolSchemaChecker;

impl McpToolSchemaChecker {
    pub fn new() -> Self {
        Self
    }

    /// Scan all files for MCP tool schema violations.
    pub fn check_mcp_tool_schema(
        &self,
        files: &[FilePath],
        _root_dir: &FilePath,
        results: &mut LintResultList,
    ) {
        for f in files {
            if !f.has_extension("py") {
                continue;
            }
            self._check_file(f, results);
        }
    }

    /// Parse a Python file and find tool registration patterns.
    fn _check_file(&self, f: &FilePath, results: &mut LintResultList) {
        let content = match std::fs::read_to_string(f.to_string()) {
            Ok(c) => c,
            Err(_) => return,
        };

        let lines: Vec<&str> = content.lines().collect();
        let mut pending_decorators: Vec<(usize, String)> = Vec::new();

        for (i, line) in lines.iter().enumerate() {
            let trimmed = line.trim();

            // Collect decorator lines
            if let Some(cap) = DECORATOR_RE.captures(trimmed) {
                let decorator_text = cap.get(1).map(|m| m.as_str()).unwrap_or("");
                pending_decorators.push((i + 1, decorator_text.to_string()));
                continue;
            }

            // When we hit a function def, check accumulated decorators
            if let Some(cap) = FUNC_DEF_RE.captures(trimmed) {
                let func_name = cap.get(1).map(|m| m.as_str()).unwrap_or("");
                let params_str = cap.get(2).map(|m| m.as_str()).unwrap_or("");

                let mut is_tool = false;
                for (_, dec_text) in &pending_decorators {
                    if self._is_tool_decorator(dec_text) {
                        is_tool = true;
                        break;
                    }
                }

                if is_tool {
                    self._check_tool_schema(
                        func_name,
                        params_str,
                        &lines,
                        i,
                        f,
                        results,
                    );
                }

                pending_decorators.clear();
            } else if !trimmed.starts_with('#') && !trimmed.is_empty() {
                // Non-decorator, non-function line resets pending decorators
                // (unless it's a blank line or comment)
                if !trimmed.is_empty() && !trimmed.starts_with('#') && !trimmed.starts_with('@') {
                    pending_decorators.clear();
                }
            }
        }
    }

    // ------------------------------------------------------------------
    // Detection
    // ------------------------------------------------------------------

    /// Determine if a decorator string represents an MCP tool registration.
    fn _is_tool_decorator(&self, decorator_text: &str) -> bool {
        for pattern in &*TOOL_DECORATOR_PATTERNS {
            if pattern.is_match(decorator_text) {
                return true;
            }
        }
        // Also check for `.tool` attribute access
        decorator_text.contains(".tool") || decorator_text == "tool"
    }

    // ------------------------------------------------------------------
    // Schema validation
    // ------------------------------------------------------------------

    /// Validate the schema associated with a tool function.
    fn _check_tool_schema(
        &self,
        func_name: &str,
        params_str: &str,
        lines: &[&str],
        func_line_idx: usize,
        f: &FilePath,
        results: &mut LintResultList,
    ) {
        self._check_docstring(func_name, lines, func_line_idx, f, results);
        self._check_parameter_types(func_name, params_str, func_line_idx, f, results);
        self._check_explicit_schemas(func_name, lines, func_line_idx, f, results);
    }

    /// Tools must have a docstring — this becomes the tool description in tools/list.
    fn _check_docstring(
        &self,
        func_name: &str,
        lines: &[&str],
        func_line_idx: usize,
        f: &FilePath,
        results: &mut LintResultList,
    ) {
        // Look for docstring in the lines following the function def
        let mut found_docstring = false;
        for line in &lines[func_line_idx + 1..] {
            let trimmed = line.trim();
            if trimmed.is_empty() {
                continue;
            }
            if DOCSTRING_RE.is_match(trimmed) {
                // Check length >= 10 chars
                let stripped = trimmed
                    .trim_start_matches(|c| c == '"' || c == '\'')
                    .trim_end_matches(|c| c == '"' || c == '\'')
                    .trim();
                if stripped.len() >= 10 {
                    found_docstring = true;
                }
            } else if trimmed.starts_with("def ") || trimmed.starts_with("class ") || trimmed.starts_with("@") {
                break;
            }
        }

        if !found_docstring {
            results.append(LintResult {
                file: f.clone(),
                line: LineNumber::new((func_line_idx as i64) + 1),
                column: ColumnNumber::new(1),
                code: ErrorCode::new("AES025").unwrap(),
                message: LintMessage::new(format!(
                    "AES025 MCP_TOOL_SCHEMA_VIOLATION: MCP tool '{}' is missing a descriptive docstring.\n\
                     WHY? The docstring becomes the tool description in tools/list response — models use it for routing.\n\
                     FIX: Add a docstring describing what this tool does, its inputs, and expected output.",
                    func_name
                )),
                source: Some(AdapterName::new("mcp_tool_schema").unwrap()),
                severity: Severity::CRITICAL,
                enclosing_scope: None,
                related_locations: LocationList::new(),
            });
        }
    }

    /// All non-self parameters on a tool function must have type annotations.
    fn _check_parameter_types(
        &self,
        func_name: &str,
        params_str: &str,
        func_line_idx: usize,
        f: &FilePath,
        results: &mut LintResultList,
    ) {
        if params_str.trim().is_empty() {
            return;
        }

        for param in params_str.split(',') {
            let param = param.trim();
            if param.is_empty() || param == "self" || param == "ctx" {
                continue;
            }
            // Strip default value if present
            let param_name = param.split('=').next().unwrap_or(param).trim();
            // Strip keyword-only marker (*, *)
            if param_name.starts_with('*') || param_name.starts_with("**") {
                continue;
            }

            // Check if it has a type annotation (contains ':')
            if !param_annotated(param_name) {
                results.append(LintResult {
                    file: f.clone(),
                    line: LineNumber::new((func_line_idx as i64) + 1),
                    column: ColumnNumber::new(1),
                    code: ErrorCode::new("AES025").unwrap(),
                    message: LintMessage::new(format!(
                        "AES025 MCP_TOOL_SCHEMA_VIOLATION: MCP tool '{}' parameter '{}' lacks a type annotation.\n\
                         WHY? Untyped parameters cannot be mapped to JSON Schema in the tools/list schema — models won't know the input format.\n\
                         FIX: Add a type annotation (e.g., str, int, FilePath, or a Pydantic model).",
                        func_name, param_name
                    )),
                    source: Some(AdapterName::new("mcp_tool_schema").unwrap()),
                    severity: Severity::CRITICAL,
                    enclosing_scope: None,
                    related_locations: LocationList::new(),
                });
            }
        }
    }

    /// Check for explicit JSON Schema dicts passed to the tool decorator.
    fn _check_explicit_schemas(
        &self,
        func_name: &str,
        lines: &[&str],
        func_line_idx: usize,
        f: &FilePath,
        results: &mut LintResultList,
    ) {
        // Look backwards from the function def to find the decorator with keyword args
        let mut i = func_line_idx;
        while i > 0 {
            i -= 1;
            let line = lines[i].trim();
            if line.starts_with('@') {
                // Check for parameters=, input_schema=, schema= keywords
                if line.contains("parameters=") || line.contains("input_schema=") || line.contains("schema=") {
                    self._validate_schema_from_line(func_name, line, f, results, (func_line_idx as i64) + 1);
                }
                break;
            }
            if !line.is_empty() && !line.starts_with('#') {
                break;
            }
        }
    }

    /// If an explicit schema dict is found in the line, validate it.
    fn _validate_schema_from_line(
        &self,
        func_name: &str,
        line: &str,
        f: &FilePath,
        results: &mut LintResultList,
        line_number: i64,
    ) {
        // Extract the schema dict portion from the decorator line
        // Look for patterns like: parameters={"type": "object", ...}
        let mut violations: Vec<String> = Vec::new();

        // Check for 'type' keyword in the dict literal
        if !line.contains("\"type\"") && !line.contains("'type'") {
            violations.push("Schema missing 'type' keyword".to_string());
        } else if !line.contains("\"properties\"") && !line.contains("'properties'") {
            violations.push("Schema has 'type' but no 'properties'".to_string());
        }

        // Check type values
        if let Some(type_val) = extract_schema_type_value(line) {
            if !JSON_SCHEMA_TYPE_VALUES.contains(type_val.as_str()) {
                violations.push(format!("Schema type='{}' is not a valid JSON Schema type", type_val));
            }
        }

        if !violations.is_empty() {
            self._report_aes025(func_name, violations, f, results, line_number);
        }
    }

    // -- Sub-checks for _validate_schema_value -----------------------------------

    /// Verify that required JSON Schema keywords are present.
    fn _check_required_keywords(&self, keys: &[String], violations: &mut Vec<String>) {
        if !keys.iter().any(|k| k == "type") {
            violations.push("Schema missing 'type' keyword".to_string());
            return;
        }
        if !keys.iter().any(|k| k == "properties") {
            violations.push("Schema has 'type' but no 'properties'".to_string());
        }
    }

    /// Ensure each property in the schema has a description.
    fn _check_property_descriptions(
        &self,
        schema_line: &str,
        keys: &[String],
        violations: &mut Vec<String>,
    ) {
        if !keys.iter().any(|k| k == "properties") {
            return;
        }
        // Simple heuristic: look for property names and check if they have descriptions
        // In a dict literal like {"type": "object", "properties": {"name": {"type": "string", "description": "..."}}}
        // We look for property sub-dicts without "description"
        let props_start = schema_line.find("\"properties\"").or_else(|| schema_line.find("'properties'"));
        if let Some(start) = props_start {
            let snippet = &schema_line[start..];
            // Count property entries that lack description
            let re = Regex::new(r#""(\w+)"\s*:\s*\{[^}]*\}"#).unwrap();
            for cap in re.captures_iter(snippet) {
                if let Some(prop_name) = cap.get(1) {
                    let prop_dict = cap.get(0).map(|m| m.as_str()).unwrap_or("");
                    if !prop_dict.contains("description") {
                        violations.push(format!(
                            "Property '{}' missing description in schema",
                            prop_name.as_str()
                        ));
                    }
                }
            }
        }
    }

    /// Validate that 'type' values are valid JSON Schema types.
    fn _check_type_values(
        &self,
        schema_line: &str,
        _keys: &[String],
        violations: &mut Vec<String>,
    ) {
        if let Some(type_val) = extract_schema_type_value(schema_line) {
            if !JSON_SCHEMA_TYPE_VALUES.contains(type_val.as_str()) {
                violations.push(format!(
                    "Schema type='{}' is not a valid JSON Schema type",
                    type_val
                ));
            }
        }
    }

    /// Append a single AES025 result to the results list.
    fn _report_aes025(
        &self,
        func_name: &str,
        violations: Vec<String>,
        f: &FilePath,
        results: &mut LintResultList,
        line: i64,
    ) {
        let detail: String = violations
            .iter()
            .map(|v| format!("  - {}", v))
            .collect::<Vec<_>>()
            .join("\n");

        results.append(LintResult {
            file: f.clone(),
            line: LineNumber::new(line),
            column: ColumnNumber::new(1),
            code: ErrorCode::new("AES025").unwrap(),
            message: LintMessage::new(format!(
                "AES025 MCP_TOOL_SCHEMA_VIOLATION: MCP tool '{}' has an invalid JSON Schema:\n{}\n\
                 WHY? MCP tools must declare valid JSON Schema so LLM clients can validate input before tool calls.\n\
                 FIX: Use a Pydantic BaseModel for tool parameters or provide a valid dict with 'type' and 'properties' keys.",
                func_name, detail
            )),
            source: Some(AdapterName::new("mcp_tool_schema").unwrap()),
            severity: Severity::CRITICAL,
            enclosing_scope: None,
            related_locations: LocationList::new(),
        });
    }
}

/// Check if a parameter string has a type annotation (contains ':').
fn param_annotated(param: &str) -> bool {
    param.contains(':')
}

/// Extract the value of a "type" key from a schema dict literal string.
fn extract_schema_type_value(line: &str) -> Option<String> {
    // Match patterns like: "type": "string" or 'type': 'object'
    let re = Regex::new(r#""type"\s*:\s*"([^"]+)""#).unwrap();
    if let Some(cap) = re.captures(line) {
        return cap.get(1).map(|m| m.as_str().to_string());
    }
    let re = Regex::new(r#"'type'\s*:\s*'([^']+)'"#).unwrap();
    if let Some(cap) = re.captures(line) {
        return cap.get(1).map(|m| m.as_str().to_string());
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_tool_decorator() {
        let checker = McpToolSchemaChecker::new();
        assert!(checker._is_tool_decorator(r#"@mcp.tool(name="foo")"#));
        assert!(checker._is_tool_decorator("@mcp.tool"));
        assert!(checker._is_tool_decorator("server.add_tool"));
        assert!(checker._is_tool_decorator("register_tool"));
        assert!(!checker._is_tool_decorator("@app.get"));
        assert!(!checker._is_tool_decorator("regular_function"));
    }

    #[test]
    fn test_extract_schema_type_value() {
        assert_eq!(
            extract_schema_type_value(r#""type": "object""#),
            Some("object".to_string())
        );
        assert_eq!(
            extract_schema_type_value(r#"'type': 'string'"#),
            Some("string".to_string())
        );
        assert_eq!(extract_schema_type_value("no type here"), None);
    }
}
